The smallest number of schema changes required to support transports extensibility

cristian-recoseanu opened the pull request:
This does not include any markdown paragraph yet to introduce the concept of extending supported transports through [NMOS Transport Parameters](https://specs.amwa.tv/nmos-parameter-registers/branches/main/transport-parameters/).

cristian-recoseanu committed:
The smallest number of schema changes required to support transports extensibility

garethsb commented on APIs/schemas/transporttype-response-schema.json:
```diff
@@ -3,14 +3,23 @@
       ]
+    },
+    {
+      "description": "Any transport registered in the parameter register",
```
Probably need to make clear needs to be a URN base, like `...:rtp`, not `...:rtp.mcast`.

```suggestion
      "description": "Any transport type URN base defined in the Transports register of the NMOS Parameter Registers",
```

garethsb commented on APIs/schemas/transporttype-response-schema.json:
```diff
@@ -3,14 +3,23 @@
+    {
+      "description": "Any transport registered in the parameter register",
+      "type": "string",
+      "pattern": "^urn:x-nmos:transport:[a-zA-Z0-9_]+$"
```
Might need a few tweaks to match the description in Use of URNs in IS-04, which defines general NMOS URNs as `<URN-base>[.<subclassification>][/<version>]` and says "the URN base is a sequence of one or more names delimited by `:` characters, beginning with the namespace `urn:x-nmos:`." Of course, unhelpfully, it doesn't go on to define the permitted characters in a "name"... except that it's clear it excludes "." and "/" as well as ":".

The name pattern should allow "-" as well as "_" I think?

So maybe this:

```suggestion
      "pattern": "^urn:x-nmos:transport:[a-zA-Z0-9_-]+(:[a-zA-Z0-9_-]+)*$"
```

If we can think of reasons we might want to allow other values later (e.g. allow subclassifications or versions, or Chinese characters in names), then maybe we'd make it _less_ precise like some other examples in the spec, i.e. just:

```suggestion
      "pattern": "^urn:x-nmos:transport:"
```

cristian-recoseanu replied:
Something worth debating in an ARG discussion.

cristian-recoseanu replied:
Worth noting for external parameter keys we use `^ext_[a-zA-Z0-9_]+$`

garethsb replied:
I considered that, but URN are not the same as property names.

garethsb replied:
Noting that IS-04 uses the simpler form everywhere, e.g.

https://github.com/AMWA-TV/is-04/blob/3e7fc65109bf95acfe0d2c30a4e46ee65144684b/APIs/schemas/device.json#L23

https://github.com/AMWA-TV/is-04/blob/3e7fc65109bf95acfe0d2c30a4e46ee65144684b/APIs/schemas/receiver_core.json#L27

However, the difference here is that we want to allow only URN base without subclassification or version.
So another simple formulation could be:

```suggestion
      "pattern": "^urn:x-nmos:transport:[^./]+$"
```

garethsb commented on APIs/schemas/transporttype-response-schema.json:
```diff
@@ -3,14 +3,23 @@
+            "urn:x-nmos:transport:websocket",
+            "urn:x-nmos:transport:mqtt"
+          ]
+        }
```
The `oneOf` is redundant here, I think?

```suggestion
      "enum": [
        "urn:x-nmos:transport:rtp",
        "urn:x-nmos:transport:dash",
        "urn:x-nmos:transport:websocket",
        "urn:x-nmos:transport:mqtt"
```

garethsb commented on APIs/schemas/sender_transport_params.json:
```diff
@@ -25,6 +26,23 @@
+        "type": "object",
+        "patternProperties": {
+          "^[a-zA-Z0-9_]+$": {
+            "type": [
+              "string",
+              "boolean",
+              "null",
+              "number"
+            ]
```
Maybe we could do:

```suggestion
            "$ref": "sender_transport_params_ext.json"
```

cristian-recoseanu replied:
Since it's a very simple set of requirements I would personally say it's better to directly mention them rather than refer to the _ext schema. Ext parameters have certain connotations which are defined in a specific markdown section [externally-defined-parameters](https://specs.amwa.tv/is-05/releases/v1.1.2/docs/Behaviour.html#externally-defined-parameters).

garethsb replied:
That's true. We try to avoid duplication but we already have third schema snippet in a couple of places so doubling it isn't so bad 😅
Whoever is next to add a newly supported data type will just have to make the changes in a few places.

garethsb commented on APIs/schemas/receiver_transport_params.json:
```diff
@@ -25,6 +26,23 @@
+        "type": "object",
+        "patternProperties": {
+          "^[a-zA-Z0-9_]+$": {
+            "type": [
+              "string",
+              "boolean",
+              "null",
+              "number"
+            ]
```
Maybe we could do:

```suggestion
            "$ref": "receiver_transport_params_ext.json"
```

cristian-recoseanu replied:
Since it's a very simple set of requirements I would personally say it's better to directly mention them rather than refer to the _ext schema. Ext parameters have certain connotations which are defined in a specific markdown section [externally-defined-parameters](https://specs.amwa.tv/is-05/releases/v1.1.2/docs/Behaviour.html#externally-defined-parameters).

garethsb replied:
Same as https://github.com/AMWA-TV/is-05/pull/177#discussion_r1869809855

garethsb commented on APIs/schemas/sender_transport_params.json:
```diff
@@ -25,6 +26,23 @@
+    {
+      "type": "array",
+      "items": {
+        "description": "Any other definition for transports registered in the parameter register",
```
```suggestion
        "description": "Transport-specific parameters for another transport defined in Transports register of the NMOS Parameter Registers",
```

garethsb commented on APIs/schemas/receiver_transport_params.json:
```diff
@@ -25,6 +26,23 @@
+    {
+      "type": "array",
+      "items": {
+        "description": "Any other definition for transports registered in the parameter register",
```
```suggestion
        "description": "Transport-specific parameters for another transport defined in Transports register of the NMOS Parameter Registers",
```

garethsb commented on APIs/schemas/constraints-schema.json:
```diff
@@ -19,6 +20,18 @@
+    {
+      "type": "array",
+      "items": {
+        "description": "Any other definition for transports registered in the parameter register",
```
```suggestion
        "description": "Constraints on transport parameters for another transport defined in Transports register of the NMOS Parameter Registers",
```

garethsb reviewed (commented):
Great! I've offered some slightly more long-winded descriptions and a couple of suggested schema tweaks!

cristian-recoseanu committed:
Update APIs/schemas/transporttype-response-schema.json

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

cristian-recoseanu committed:
Update APIs/schemas/transporttype-response-schema.json

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

cristian-recoseanu committed:
Update APIs/schemas/sender_transport_params.json

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

cristian-recoseanu committed:
Update APIs/schemas/receiver_transport_params.json

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

cristian-recoseanu committed:
Update APIs/schemas/constraints-schema.json

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

alabou commented:
transporttype-response-schema.json must allow vendor defined transports as per https://specs.amwa.tv/nmos-parameter-registers/branches/main/transports "Manufacturers MAY use their own namespaces to indicate transports which are not currently defined within the NMOS namespace."

garethsb committed:
Allow vendor-specific transport types

cristian-recoseanu commented on APIs/schemas/transporttype-response-schema.json:
```diff
@@ -13,9 +13,15 @@
       "pattern": "^urn:x-nmos:transport:[a-zA-Z0-9_]+$"
+    },
+    {
+      "description": "Any manufacturer-specific transport type URN base defined in the Transports register of the NMOS Parameter Registers",
```
Do we expect manufacturer namespaces to be registered in the transports register? I feel this whole area is thin in terms of interoperability. What would a controller be expected to do when it encounters a manufacturer namespace?

alabou replied:
I would assume that the controller then check if it knows about this transport and then use the vendor specific schemas about such transport. There may be an issue if multiple vendors claim the same namespace ...

cristian-recoseanu committed:
Update APIs/schemas/transporttype-response-schema.json

Suggested change

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

garethsb committed:
Simplify the transport params and constraints schemas since an instance that is valid per any of the transport-specific schemas is also valid per the general schema
Implementations (devices, clients, test suites) may use the transport-specific schemas as an additional validation step

garethsb committed:
Simplify the transport type, transport params, and constraints, schemas
* An instance that is valid per any of the transport-specific schemas is also valid per the general schema
* Implementations (devices, clients, test suites) may use the transport-specific schemas as an additional validation step

cristian-recoseanu committed:
Merge branch 'publish-transports-extensibility' of https://github.com/AMWA-TV/is-05 into publish-transports-extensibility

cristian-recoseanu committed:
Make description of schemas more agnostic of where the transport lives

cristian-recoseanu committed:
Add transports parameter register references to relevant sections

peterbrightwell reviewed (approved):
Reviewed and approved on ARG call 2025-03-19
